options/persistent.go

```diff
@@ -0,0 +1,101 @@
+package options
+
+import (
+	"github.com/EventStore/EventStore-Client-Go/client/filtering"
+	"github.com/EventStore/EventStore-Client-Go/persistent"
+	"github.com/EventStore/EventStore-Client-Go/stream_position"
+)
+
+type PersistentStreamSubscriptionOptions struct {
+	settings persistent.SubscriptionSettings
+	position stream_position.StreamPosition
+}
+
+func PersistentStreamSubscriptionOptionsDefault() PersistentStreamSubscriptionOptions {
+	return PersistentStreamSubscriptionOptions{
+		settings: persistent.SubscriptionSettingsDefault(),
+		position: stream_position.End(),
+	}
+}
+
+func (opts PersistentStreamSubscriptionOptions) Settings(settings persistent.SubscriptionSettings) PersistentStreamSubscriptionOptions {
+	opts.settings = settings
+	return opts
+}
+
+func (opts PersistentStreamSubscriptionOptions) Position(position stream_position.StreamPosition) PersistentStreamSubscriptionOptions {
```

Comment:
Setters like this will work if the "self" object is being passed by value and not reference. It will set the field on the copy of `opts` which disappears at the end of this method.

```go
func TestPersistentAllSubscriptionOptionsDefaultPosition(t *testing.T) {
	d := PersistentAllSubscriptionOptionsDefault()
	if d.GetPosition() != stream_position.End() {
		t.Errorf("failure: stream position not set to End()")  // not called
	}
	d.Position(stream_position.Start())
	if d.GetPosition() != stream_position.Start() {
		t.Errorf("failure: stream position not set to Start()")  // is called
	}
}
```

Reply:
How embarrassing, I missed that this returns this is returning the self value. My guess is that the Go compiler probably will just move these values in most cases so it won't be inefficient.

However I agree with roblaszczak when he mentions it [might be easier to make these mutable](https://github.com/EventStore/EventStore-Client-Go/pull/65#discussion_r715084177).